### ISSUE TITLE
Large JSON import partially fails.

### ISSUE THREAD

**Issue by luissantosHCIT** [medplum/medplum#5498]
Good morning,

## Intro

I recently synchronized my fork of the project with ~107 commits. I think my initial fork was pre-October.
Since the database version was updated, I redid my test database.
In the process, I attempted to re-import a 57mb JSON bundle via the Medplum App.

This bundle is a Synthea generated bundle for R4 which had succeeded in the past. I just needed to set the following setting in the server's config.

- "maxJsonSize": "100mb"

I updated my copy of the repo due to a need to investigate another issue trying to request a chained search with _has.
Once I get my test data back, I will try to see if I need to open another ticket for the search issue.

## Error

On the latest version of the main branch, I now get this in the server terminal output.

> {"level":"WARN","timestamp":"2024-11-05T13:37:17.324Z","msg":"Unknown profile referenced","resource":"Encounter/29452cb1-7c5e-431f-805f-8c07d174e36c","url":"http://hl7.org/fhir/us/core/StructureDefinition/us-core-encounter","requestId":"82f28d8c-48db-40ef-8b8d-e76aaef7e310","traceId":"7d201994-43c3-4ad7-92f4-f6b58f387f74"}
{"level":"WARN","timestamp":"2024-11-05T13:37:17.353Z","msg":"Unknown profile referenced","resource":"DiagnosticReport/2b1242a8-a904-40cf-bafe-008374a978fe","url":"http://hl7.org/fhir/us/core/StructureDefinition/us-core-diagnosticreport-lab","requestId":"82f28d8c-48db-40ef-8b8d-e76aaef7e310","traceId":"7d201994-43c3-4ad7-92f4-f6b58f387f74"}
{"level":"WARN","timestamp":"2024-11-05T13:37:17.369Z","msg":"Unknown profile referenced","resource":"DiagnosticReport/41448fa1-eedd-4cd6-9c37-ca6de4dd9731","url":"http://hl7.org/fhir/us/core/StructureDefinition/us-core-diagnosticreport-note","requestId":"82f28d8c-48db-40ef-8b8d-e76aaef7e310","traceId":"7d201994-43c3-4ad7-92f4-f6b58f387f74"}
{"level":"WARN","timestamp":"2024-11-05T13:37:17.388Z","msg":"Unknown profile referenced","resource":"DocumentReference/048180b1-165a-4ce9-aa0c-900ea1e2662d","url":"http://hl7.org/fhir/us/core/StructureDefinition/us-core-documentreference","requestId":"82f28d8c-48db-40ef-8b8d-e76aaef7e310","traceId":"7d201994-43c3-4ad7-92f4-f6b58f387f74"}
{"level":"WARN","timestamp":"2024-11-05T13:37:17.412Z","msg":"Validator warning: Invalid reference: got \"#referral\", expected \"http://hl7.org/fhir/StructureDefinition/ServiceRequest\"","project":"ed3ec4ff-0af6-45f7-a6cc-8138f190ad8b","issue":{"severity":"warning","code":"structure","details":{"text":"Invalid reference: got \"#referral\", expected \"http://hl7.org/fhir/StructureDefinition/ServiceRequest\""},"expression":["ExplanationOfBenefit.referral"]},"requestId":"82f28d8c-48db-40ef-8b8d-e76aaef7e310","traceId":"7d201994-43c3-4ad7-92f4-f6b58f387f74"}
{"level":"WARN","timestamp":"2024-11-05T13:37:17.412Z","msg":"Validator warning: Invalid reference: got \"#coverage\", expected \"http://hl7.org/fhir/StructureDefinition/Coverage\"","project":"ed3ec4ff-0af6-45f7-a6cc-8138f190ad8b","issue":{"severity":"warning","code":"structure","details":{"text":"Invalid reference: got \"#coverage\", expected \"http://hl7.org/fhir/StructureDefinition/Coverage\""},"expression":["ExplanationOfBenefit.insurance[0].coverage"]},"requestId":"82f28d8c-48db-40ef-8b8d-e76aaef7e310","traceId":"7d201994-43c3-4ad7-92f4-f6b58f387f74"}
{"level":"WARN","timestamp":"2024-11-05T13:37:17.978Z","msg":"Unknown profile referenced","resource":"Provenance/300c5f9a-ec02-469e-9311-bc95da1d0160","url":"http://hl7.org/fhir/us/core/StructureDefinition/us-core-provenance","requestId":"82f28d8c-48db-40ef-8b8d-e76aaef7e310","traceId":"7d201994-43c3-4ad7-92f4-f6b58f387f74"}
{"level":"ERROR","timestamp":"2024-11-05T13:37:18.610Z","msg":"Database error","error":"bind message has 797 parameter formats but 0 parameters","stack":"error: bind message has 797 parameter formats but 0 parameters\n    at /home/lsantos/WebstormProjects/medplum/node_modules/pg/lib/client.js:535:17\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at SqlBuilder.execute (/home/lsantos/WebstormProjects/medplum/packages/server/src/fhir/sql.ts:419:22)\n    at ReferenceTable.insertValuesForResource (/home/lsantos/WebstormProjects/medplum/packages/server/src/fhir/lookups/reference.ts:57:5)\n    at ReferenceTable.indexResource (/home/lsantos/WebstormProjects/medplum/packages/server/src/fhir/lookups/reference.ts:35:5)\n    at async Promise.all (index 4)\n    at Repository.writeLookupTables (/home/lsantos/WebstormProjects/medplum/packages/server/src/fhir/repo.ts:1642:5)\n    at /home/lsantos/WebstormProjects/medplum/packages/server/src/fhir/repo.ts:841:7\n    at Repository.withTransaction (/home/lsantos/WebstormProjects/medplum/packages/server/src/fhir/repo.ts:2151:22)\n    at Repository.writeToDatabase (/home/lsantos/WebstormProjects/medplum/packages/server/src/fhir/repo.ts:838:5)","code":"08P01","requestId":"82f28d8c-48db-40ef-8b8d-e76aaef7e310","traceId":"7d201994-43c3-4ad7-92f4-f6b58f387f74"}
{"level":"WARN","timestamp":"2024-11-05T13:37:18.613Z","msg":"Error processing transaction Bundle","errors":69,"project":"ed3ec4ff-0af6-45f7-a6cc-8138f190ad8b","requestId":"82f28d8c-48db-40ef-8b8d-e76aaef7e310","traceId":"7d201994-43c3-4ad7-92f4-f6b58f387f74"}

The input test Bundle is 
[Rozella39_Mertz280_ef167059-cef0-12c4-49db-993ca3a20c01.zip](https://github.com/user-attachments/files/17633524/Rozella39_Mertz280_ef167059-cef0-12c4-49db-993ca3a20c01.zip)

## Miscellaneous

The new server config has the option. 

- "maxBatchSize": "50mb"

Mentioning it in case recent changes expose an unexpected issue.
I tried changing the batch size to 100mb without luck.

I am running the server in "dev" mode using the new Postgres 16 database container.

## Conclusion
A large JSON Bundle seems to have an error on import. I will try other test datasets for my own testing. I have not narrowed down the issue to the exact code area, but I will try to inform you of any further findings I make.

Let me know if I need to provide more information.
Thank you for Medplum!

**Comment by mattwiller** [comment]
Hi @luissantosHCIT — thanks for the report!  The `maxBatchSize` configuration option sets the maximum size for [asynchronous batch requests](https://www.medplum.com/docs/fhir-datastore/fhir-batch-requests#asynchronous-batch-requests), which I would strongly recommend using for batches larger than a megabyte or so.

Separately, I can start looking into the database error you're seeing: it would appear that some resource in the Bundle might have more references than we can insert into the lookup table in one `INSERT` statement

**Comment by luissantosHCIT** [comment]
Ah, I see.

Which source and line number should I look into to understand where you identified the issue?

I think I will try to make some time on Friday to look deeper at the lower level APIs.
No promises, but I am always looking for ways to contribute and I am still learning the codebase.

**Comment by mattwiller** [comment]
I noticed [these lines]((https://github.com/medplum/medplum/blob/70a28416da7baef8f9d162659288785e15ff52a7/packages/server/src/fhir/lookups/reference.ts#L35-L57) in the stack trace of the error you posted, which is where references contained in a resource are inserted into a lookup table

**Comment by mattwiller** [comment]
@luissantosHCIT I've been trying to reproduce the `bind message has 797 parameter formats but 0 parameters` error you saw, but so far haven't been able to get that failure to show up.  Uploading the Bundle you linked produces mostly errors for me, since I don't have the corresponding Practitioner and Organization resources to resolve conditional references in the Bundle.  If possible, could you let me know which resource specifically is failing on your machine?

**Comment by luissantosHCIT** [comment]
This is the organization and practitioner resources I had imported in my dev environment.
[practitionerInformation1637345232350.zip](https://github.com/user-attachments/files/17733533/practitionerInformation1637345232350.zip)
[hospitalInformation1637345232350.zip](https://github.com/user-attachments/files/17733537/hospitalInformation1637345232350.zip)

I used this dataset as a testing dataset.
https://mitre.box.com/shared/static/ylzmiichhvtw1igr4ck6q32i5b333nqs.zip from https://synthea.mitre.org/downloads.
Maybe the issue is specific to something I did when importing data into the development db.

Unfortunately, I had to write some automation for a different project last Friday so I had to postpone doing a deep dive until next week. I'll return to this next week and see if I can figure out if there is something that could be handled more gracefully or if it is purely a me problem. I also need to check if a recent release may have fixed the issue altogether.

Thank you for looking into this issue.

**Comment by mattwiller** [comment]
@luissantosHCIT Thanks to the additional context you provided, I was able to reproduce the issue.  It looks like there's a Provenance resource in the main Bundle with over 20k references in it, resulting in an `INSERT` statement with 66,333 placeholder formats.  This overflows the 16-bit counter Postgres must be using internally, and gets erroneously reported as 797 in the error message.

I believe the correct fix here would be to split that `INSERT`s into batches of ~10,000 or so references: I'll start putting that PR together

**Comment by luissantosHCIT** [comment]
Interesting. I wonder if it is a side effect of moving to the newer Postgres 16 database engine. The error message did not show up previously until I nuked my dev db and synchronized my working fork of this repo.

At any rate, that is a pretty interesting find!

**Comment by luissantosHCIT** [comment]
Adding a little documentation here that may be helpful and relevant to this thread.

I found a limits documentation page in Postgres and they have a 65K limit on query parameters. 
https://www.postgresql.org/docs/16/limits.html

![image](https://github.com/user-attachments/assets/dfdf83c7-8b21-4b1e-9830-e5a11caa04ed)

I am not sure why it did not come up the first time I imported this dataset (pre-16), but seems like they had this limit all along.

**Comment by mattwiller** [comment]
The 64k limit on parameters does seem to be a long-standing limit.  If you only recently synchronized your fork, I wonder if previously you were operating on a commit prior to [adding the Reference lookup tables](https://github.com/medplum/medplum/commit/ac2e8f96bf0c46bbebbacbdf005c0e26a7bd3ae8), which is where this error was occurring

**Comment by luissantosHCIT** [comment]
No, I don't think so. I learned about the project around late April, then I had to do other stuff, and finally came back to studying it near September. I did not start importing the trouble bundle until more recently, so maybe I did not notice or the error message was hidden from all other import messages. When I synched my April copy to the prior to last release, I was more focused on getting the large dataset imported because it had so many resources defined. It's hard to say what exactly made the import issue more visible (if it was hidden).

Anyways, thank you!
